### PR TITLE
New case of larger_buffer_with_none_cache_mode

### PR DESCRIPTION
**Pull request by hellohellenmao**
1. Create 100M source image via 'dd'.
2. Convert the source image and check whether the maxim
   result of pwrite/pwrite64 system call is 2M.

ID: 1784696
Signed-off-by: Tingting Mao <timao@redhat.com>

### CODE REVIEW

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.larger_buffer_with_none_cache_mode: PASS

**Comment by hellohellenmao** [comment]
@lolyu @yhong Help to take review, please.

File: qemu/tests/larger_buffer_with_none_cache_mode.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    logging.debug("Convert from %s to %s with cache mode none, strace log: %s."
+                  % (src_filename, tgt_image, strace_output_file))
+    with strace(source, strace_events, strace_output_file, trace_child=True):
+        fail_on((process.CmdError,))(source.convert)(
+            params.object_params(src_image), root_dir, cache_mode="none")
+
+    logging.debug("Check whether the max size of %s syscall is 2M in %s." %
+                  (strace_events, strace_output_file))
+    with open(strace_output_file) as fd:
+        if "= 2097152" not in fd.read():
```
Comment (lolyu) [inline qemu/tests/larger_buffer_with_none_cache_mode.py]: the check is inaccuracy, the buffer size for `pwrite` will be at max 2M.

File: qemu/tests/larger_buffer_with_none_cache_mode.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    src_image = params["images"]
+    tgt_image = params["convert_target"]
+    root_dir = data_dir.get_data_dir()
+    source = QemuImg(params.object_params(src_image), root_dir, src_image)
+    strace_event = params["strace_event"]
+    strace_events = strace_event.split()
+    strace_output_file = os.path.join(test.debugdir,
+                                      "convert_with_none.log")
+    src_filename = source.image_filename
+    process.run("dd if=/dev/urandom of=%s bs=1M count=100" % src_filename)
```
[the remark this answers is not in the document]
Comment (hellohellenmao) [reply qemu/tests/larger_buffer_with_none_cache_mode.py]: `create_with_dd` is for 'zero' images creation. But this case need real data in it. Can not use it.

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.larger_buffer_with_none_cache_mode: PASS (3.00 s)

File: qemu/tests/larger_buffer_with_none_cache_mode.py (hunk 1 above)
Comment (hellohellenmao) [reply qemu/tests/larger_buffer_with_none_cache_mode.py]: Updated. Thanks. Help to review again, please.

File: qemu/tests/larger_buffer_with_none_cache_mode.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                  (strace_event, strace_output_file))
+    buffer_size_lst = []
+    with open(strace_output_file) as fd:
+        for line in fd.readlines():
+            if strace_event in line:
+                buffer_size_lst.append(int(line.split()[-1]))
+    max_buffer = max(buffer_size_lst)
+    if max_buffer != 2097152:
+        test.fail("The max size of '%s' is %s not 2M, check '%s' please."
+                  % (strace_event, str(max_buffer), strace_output_file))
```
Comment (YongxueHong) [inline qemu/tests/larger_buffer_with_none_cache_mode.py]: How about ?

```
        for line in fd.readlines():
            if "%s" % strace_events in line and int(line.split()[-1]) == 2097152:
                    break
        else:
            test.fail("The maxim size of '%s' is not 2M, check '%s' please."
                      % (strace_event, strace_output_file))
```

You don't need to create a new list and call the `max` built-in function and reduce the times of calling in the `for` loop.

Comment (hellohellenmao) [reply qemu/tests/larger_buffer_with_none_cache_mode.py]: Updated, thanks.

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.larger_buffer_with_none_cache_mode: PASS (3.08 s)

File: qemu/tests/larger_buffer_with_none_cache_mode.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+                  % (src_filename, tgt_image, strace_output_file))
+    with strace(source, strace_events, strace_output_file, trace_child=True):
+        fail_on((process.CmdError,))(source.convert)(
+            params.object_params(src_image), root_dir, cache_mode="none")
+
+    logging.debug("Check whether the max size of %s syscall is 2M in %s." %
+                  (strace_event, strace_output_file))
+    with open(strace_output_file) as fd:
+        for line in fd.readlines():
+            if strace_event in line and int(line.split()[-1]) == 2097152:
```
Comment (lolyu) [inline qemu/tests/larger_buffer_with_none_cache_mode.py]: no need to check if `strace_event` is included in line.

Comment (hellohellenmao) [reply qemu/tests/larger_buffer_with_none_cache_mode.py]: Yeah, you're right. Updated, thanks.

**Comment by hellohellenmao** [comment]
@lolyu Help to review again. Merge if every thing is okay to you, please.

File: qemu/tests/larger_buffer_with_none_cache_mode.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    tgt_image = params["convert_target"]
+    root_dir = data_dir.get_data_dir()
+    source = QemuImg(params.object_params(src_image), root_dir, src_image)
+    strace_event = params["strace_event"]
+    strace_events = strace_event.split()
+    strace_output_file = os.path.join(test.debugdir,
+                                      "convert_with_none.log")
+    src_filename = source.image_filename
+    process.run("dd if=/dev/urandom of=%s bs=1M count=100" % src_filename)
+    logging.debug("Convert from %s to %s with cache mode none, strace log: %s."
```
Comment (lolyu) [inline qemu/tests/larger_buffer_with_none_cache_mode.py]:
```python
logging.debug("Convert from %s to %s with cache mode none, strace log: %s.", src_filename, tgt_image, strace_output_file)
```

File: qemu/tests/larger_buffer_with_none_cache_mode.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+                                      "convert_with_none.log")
+    src_filename = source.image_filename
+    process.run("dd if=/dev/urandom of=%s bs=1M count=100" % src_filename)
+    logging.debug("Convert from %s to %s with cache mode none, strace log: "
+                  "%s.", src_filename, tgt_image, strace_output_file)
+    with strace(source, strace_events, strace_output_file, trace_child=True):
+        fail_on((process.CmdError,))(source.convert)(
+            params.object_params(src_image), root_dir, cache_mode="none")
+
+    logging.debug("Check whether the max size of %s syscall is 2M in %s." %
```
Comment (lolyu) [inline qemu/tests/larger_buffer_with_none_cache_mode.py]: same as above.

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m8.u2.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.larger_buffer_with_none_cache_mode: PASS (15.41 s)

**Comment by hellohellenmao** [comment]
(1/1) Host_RHEL.m8.u2.raw.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.x86_64.io-github-autotest-qemu.larger_buffer_with_none_cache_mode: PASS (15.51 s)

**Pull request merged**